### PR TITLE
Simplify traversal

### PR DESCRIPTION
**Pull request by thet**
by @petri 
Opening this PR to be able to discuss it right here.

### CODE REVIEW

**Comment by thet** [comment]
@jenkins-plone-org please run jobs

File: src/plone/rest/traverse.py [hunk 1]
```diff
@@ -14,35 +12,22 @@
...
-            return obj
-
-        # Do not handle view namespace
-        if "@@" in request["PATH_INFO"] or "++view++" in request["PATH_INFO"]:
-            return obj
+        obj = super(RESTTraverse, self).publishTraverse(request, name)
 
         # Wrap object to ensure we handle further traversal
-        return RESTWrapper(obj, request)
+        if IContentish.providedBy(obj) and not (
```
Comment (petri) [inline src/plone/rest/traverse.py:25]: I am not sure why we check for IContentish here, rather than ITraversable? I tried that and it seems to solve https://github.com/plone/plone.restapi/issues/680 and https://github.com/plone/plone.app.theming/issues/165 whilst still passing at least `plone.rest` tests.

Comment (tisto) [reply src/plone/rest/traverse.py:25]: Interesting. I'd guess that IContentish is always ITraversable. IContentish excludes the SiteRoot, not sure if that is traversable. Things are complex and different people worked on different parts that depend on plone.rest, so I'd like them to weight in as well. cc @buchi @jaroel @sneridagh

Comment (jaroel) [reply src/plone/rest/traverse.py:25]: I actively remove `IContentish` from `Plone Site` in https://github.com/plone/Products.CMFPlone/pull/2612/files#diff-860754a57fed2533e4aae6aab34a27caR238-R242 . I did not check how that will affect `plone.rest` stuff, but so far it seemed to work.
My gut reaction is that `ITraversable` is correct here.

**Comment by petri** [comment]
@jenkins-plone-org please run jobs

**Comment by jensens** [comment]
@jenkins-plone-org please run jobs

**Comment by petschki** [comment]
@jenkins-plone-org please run jobs

**Comment by petschki** [comment]
@jenkins-plone-org please run jobs

**Comment by petschki** [comment]
@jenkins-plone-org please run jobs

**Comment by petschki** [comment]
@jenkins-plone-org please run jobs

**Comment by tisto** [comment]
@thet @petri @petschki could you elaborate on the background of this simplification? What's the actual use case here and the reasoning behind this PR?

The traversal mechanism has been subject to multiple refactorings in the past and it needs to cover multiple edge cases that we can not always put into proper unit tests. Changing this requires careful consideration and extensive testing and review. I'd like to understand the reasoning behind this for a proper tradeoff of the risks and chances here.

**Comment by petschki** [comment]
I'll try my best...

OLD strategy:

Try Zope's DefaultPublishTraverse first an catch it's `KeyError` as trigger to do check for REST service or BrowserView and if nothing found re-raise KeyError to go back to the DefaulPublishTraverse again. The downside of this is, that the special `isinstance` check for VirtualHostMonster (and possibly other checks) have to be included in plone.rest traversal.

NEW strategy:

1. Check for `IService` -> we can return this early
2. Go with the DefaultPublishTraverse -> this raises KeyError (eg. if traversed name not found) and follows the default zope NotFound traversal mechanism
3. Check for contentish objects -> wrap them with the RESTWrapper
4. Last return object if it's something we do not cover (eg. VirtualHostMonster)

So this does exactly the same as before with the benefit of getting rid of the VirtualHostMonster (and possibly other objects) check. If there are missing special traversal objects/paths, they were missing before too. But they can be easily extented after step 3. and before step 4.

hope this helps.

**Comment by petschki** [comment]
NOTE: this branch is now on a Plone4.3/collective.exportimport migration project of us and works like a charm. The linage subsites delivered with VirtualHostMonster returns now plone.restapi JSON responses. Before there was this error:

```
"Resource not found: https://www.klaus.at/Plone/klaus/virtual_hosting"
```

/cc @thet I think you also had this one?

**Comment by petschki** [comment]
btw. the Jenkins jobs are flapping randomly ... yesterday it was https://jenkins.plone.org/job/pull-request-5.2-3.8/  today it's https://jenkins.plone.org/job/pull-request-5.2-2.7/ but I didn't change anything ... has anybody a clue on this?

/cc @jensens @rodfersou

**Comment by tisto** [comment]
Thanks for the clarification @petschki.

@buchi @sneridagh could you have a look at this PR?

**Comment by petri** [comment]
If someone wants to get some more background on this PR, I added a link to the discussion (see thet's original post at the beginning of this PR). It in turn has pointers to some earlier discussions. One thing worth mentioning is that the PR should also let plone.app.theming work properly again... afaik it has been partially broken ever since plone.rest was included with plone core.

**Comment by petri** [comment]
@jenkins-plone-org please run jobs

File: src/plone/rest/traverse.py (hunk 1 above)
Comment (jaroel) [reply src/plone/rest/traverse.py:25]: We're adding IContentish back to PloneSite.
I don't think this changes my position on ITraversable, but it probably will affect other stuff.

Is this whole branch/approach still relevant actually?
